File: macos/API/AudioError.swift
import Foundation

/// AudioCaptureError - Comprehensive error types for the audio capture library
public enum AudioCaptureError: LocalizedError {
    
    // MARK: - Device Errors
    case deviceNotFound(String)
    case deviceEnumerationFailed
    case deviceSelectionFailed
    case devicePropertyReadFailed
    case invalidDevice(String)
    case deviceDisconnected(String)
    case deviceInUse(String)
    
    // MARK: - Permission Errors
    case permissionDenied
    case screenRecordingPermissionRequired
    case microphonePermissionRequired
    
    // MARK: - Session Errors
    case sessionNotFound(UUID)
    case invalidState(String)
    case sessionAlreadyActive
    case sessionNotActive
    case sessionStartFailed(String)
    
    // MARK: - Format Errors
    case unsupportedFormat(String)
    case formatConversionFailed(String)
    case formatMismatch(String)
    case formatNegotiationFailed
    
    // MARK: - Output Errors
    case outputNotConfigured
    case outputConfigurationFailed(String)
    case outputProcessingFailed(String)
    case fileWriteFailed(String)
    case streamingFailed(String)
    
    // MARK: - Buffer Errors
    case bufferAllocationFailed
    case bufferOverflow
    case bufferUnderrun
    case invalidBufferSize
    
    // MARK: - System Errors
    case systemResourcesExhausted
    case memoryAllocationFailed
    case audioEngineStartFailed(String)
    case unknownError(String)
    
    // MARK: - Network Errors (for future streaming)
    case networkConnectionFailed(String)
    case streamingProtocolError(String)
    
    // MARK: - LocalizedError Implementation
    
    public var errorDescription: String? {
        switch self {
        // Device Errors
        case .deviceNotFound(let name):
            return "Audio device '\(name)' not found"
        case .deviceEnumerationFailed:
            return "Failed to enumerate audio devices"
        case .deviceSelectionFailed:
            return "Failed to select audio device"
        case .devicePropertyReadFailed:
            return "Failed to read device properties"
        case .invalidDevice(let reason):
            return "Invalid device: \(reason)"
        case .deviceDisconnected(let name):
            return "Device '\(name)' was disconnected"
        case .deviceInUse(let name):
            return "Device '\(name)' is already in use"
            
        // Permission Errors
        case .permissionDenied:
            return "Permission denied"
        case .screenRecordingPermissionRequired:
            return "Screen Recording permission is required. Please grant permission in System Settings > Privacy & Security > Screen Recording"
        case .microphonePermissionRequired:
            return "Microphone permission is required. Please grant permission in System Settings > Privacy & Security > Microphone"
            
        // Session Errors
        case .sessionNotFound(let id):
            return "Session with ID \(id) not found"
        case .invalidState(let message):
            return "Invalid session state: \(message)"
        case .sessionAlreadyActive:
            return "Session is already active"
        case .sessionNotActive:
            return "Session is not active"
        case .sessionStartFailed(let reason):
            return "Failed to start session: \(reason)"
            
        // Format Errors
        case .unsupportedFormat(let format):
            return "Unsupported audio format: \(format)"
        case .formatConversionFailed(let reason):
            return "Format conversion failed: \(reason)"
        case .formatMismatch(let details):
            return "Audio format mismatch: \(details)"
        case .formatNegotiationFailed:
            return "Failed to negotiate compatible audio format"
            
        // Output Errors
        case .outputNotConfigured:
            return "Output is not configured"
        case .outputConfigurationFailed(let reason):
            return "Output configuration failed: \(reason)"
        case .outputProcessingFailed(let reason):
            return "Output processing failed: \(reason)"
        case .fileWriteFailed(let reason):
            return "File write failed: \(reason)"
        case .streamingFailed(let reason):
            return "Streaming failed: \(reason)"
            
        // Buffer Errors
        case .bufferAllocationFailed:
            return "Failed to allocate audio buffer"
        case .bufferOverflow:
            return "Audio buffer overflow"
        case .bufferUnderrun:
            return "Audio buffer underrun"
        case .invalidBufferSize:
            return "Invalid buffer size"
            
        // System Errors
        case .systemResourcesExhausted:
            return "System resources exhausted"
        case .memoryAllocationFailed:
            return "Memory allocation failed"
        case .audioEngineStartFailed(let reason):
            return "Audio engine failed to start: \(reason)"
        case .unknownError(let message):
            return "Unknown error: \(message)"
            
        // Network Errors
        case .networkConnectionFailed(let reason):
            return "Network connection failed: \(reason)"
        case .streamingProtocolError(let reason):
            return "Streaming protocol error: \(reason)"
        }
    }
    
    public var failureReason: String? {
        switch self {
        case .screenRecordingPermissionRequired:
            return "The app needs access to system audio"
        case .microphonePermissionRequired:
            return "The app needs access to microphone input"
        case .systemResourcesExhausted:
            return "Not enough CPU or memory available"
        case .bufferOverflow:
            return "Audio processing can't keep up with input rate"
        case .bufferUnderrun:
            return "Audio input is not providing data fast enough"
        default:
            return nil
        }
    }
    
    public var recoverySuggestion: String? {
        switch self {
        case .screenRecordingPermissionRequired:
            return "Open System Settings and grant Screen Recording permission to this app"
        case .microphonePermissionRequired:
            return "Open System Settings and grant Microphone permission to this app"
        case .deviceDisconnected:
            return "Reconnect the audio device or select a different device"
        case .systemResourcesExhausted:
            return "Close other applications to free up system resources"
        case .bufferOverflow:
            return "Try reducing the audio quality or closing other applications"
        case .formatMismatch:
            return "Check that all audio components are using compatible formats"
        default:
            return nil
        }
    }
    
    public var helpAnchor: String? {
        switch self {
        case .screenRecordingPermissionRequired, .microphonePermissionRequired:
            return "permissions"
        case .deviceNotFound, .deviceDisconnected, .deviceInUse:
            return "devices"
        case .unsupportedFormat, .formatConversionFailed, .formatMismatch:
            return "formats"
        case .bufferOverflow, .bufferUnderrun:
            return "performance"
        default:
            return nil
        }
    }
}

/// Error recovery strategies
public enum ErrorRecoveryStrategy {
    case retry(maxAttempts: Int, delay: TimeInterval)
    case fallback(action: () async throws -> Void)
    case ignore
    case fail
}

/// Error handler with recovery strategies
public struct ErrorHandler {
    
    /// Handle error with recovery strategy
    public static func handle(
        _ error: Error,
        strategy: ErrorRecoveryStrategy = .fail
    ) async throws {
        print("Error occurred: \(error.localizedDescription)")
        
        switch strategy {
        case .retry(let maxAttempts, let delay):
            // Retry logic would be implemented by caller
            throw error
            
        case .fallback(let action):
            try await action()
            
        case .ignore:
            // Log and continue
            print("Ignoring error: \(error)")
            
        case .fail:
            throw error
        }
    }
    
    /// Get suggested recovery strategy for error
    public static func suggestedStrategy(for error: Error) -> ErrorRecoveryStrategy {
        if let captureError = error as? AudioCaptureError {
            switch captureError {
            case .deviceDisconnected:
                return .retry(maxAttempts: 3, delay: 1.0)
            case .bufferOverflow, .bufferUnderrun:
                return .ignore
            case .sessionStartFailed:
                return .retry(maxAttempts: 2, delay: 0.5)
            default:
                return .fail
            }
        }
        
        return .fail
    }
}

/// Error context for detailed debugging
public struct ErrorContext {
    public let error: Error
    public let timestamp: Date
    public let sessionId: UUID?
    public let operation: String
    public let additionalInfo: [String: Any]
    
    public init(
        error: Error,
        sessionId: UUID? = nil,
        operation: String,
        additionalInfo: [String: Any] = [:]
    ) {
        self.error = error
        self.timestamp = Date()
        self.sessionId = sessionId
        self.operation = operation
        self.additionalInfo = additionalInfo
    }
    
    /// Create detailed error report
    public func report() -> String {
        var report = """
        Audio Capture Error Report
        ========================
        Timestamp: \(timestamp)
        Operation: \(operation)
        Error: \(error.localizedDescription)
        """
        
        if let sessionId = sessionId {
            report += "\nSession ID: \(sessionId)"
        }
        
        if !additionalInfo.isEmpty {
            report += "\n\nAdditional Information:"
            for (key, value) in additionalInfo {
                report += "\n  \(key): \(value)"
            }
        }
        
        if let captureError = error as? AudioCaptureError,
           let suggestion = captureError.recoverySuggestion {
            report += "\n\nRecovery Suggestion: \(suggestion)"
        }
        
        return report
    }
}

File: windows/API/AudioError.py
"""
AudioCaptureError - Comprehensive error types for the Windows audio capture library
"""

from datetime import datetime
from typing import Dict, Any, Optional, Callable
from enum import Enum, auto
import uuid
import time


class ErrorRecoveryStrategy(Enum):
    """Error recovery strategies"""
    RETRY = auto()
    FALLBACK = auto()
    IGNORE = auto()
    FAIL = auto()


class AudioCaptureError(Exception):
    """Base exception class for audio capture errors"""
    
    def __init__(self, message: str, details: Optional[str] = None):
        self.message = message
        self.details = details
        super().__init__(self.message)
    
    @property
    def failure_reason(self) -> Optional[str]:
        """Get the failure reason for this error"""
        return None
    
    @property
    def recovery_suggestion(self) -> Optional[str]:
        """Get recovery suggestion for this error"""
        return None
    
    @property
    def help_anchor(self) -> Optional[str]:
        """Get help documentation anchor for this error"""
        return None


# Device Errors
class DeviceNotFoundError(AudioCaptureError):
    """Audio device not found"""
    def __init__(self, device_name: str):
        super().__init__(f"Audio device '{device_name}' not found")
        self.device_name = device_name
    
    @property
    def help_anchor(self) -> str:
        return "devices"


class DeviceEnumerationFailedError(AudioCaptureError):
    """Failed to enumerate audio devices"""
    def __init__(self):
        super().__init__("Failed to enumerate audio devices")
    
    @property
    def help_anchor(self) -> str:
        return "devices"


class DeviceSelectionFailedError(AudioCaptureError):
    """Failed to select audio device"""
    def __init__(self):
        super().__init__("Failed to select audio device")
    
    @property
    def help_anchor(self) -> str:
        return "devices"


class DevicePropertyReadFailedError(AudioCaptureError):
    """Failed to read device properties"""
    def __init__(self):
        super().__init__("Failed to read device properties")


class InvalidDeviceError(AudioCaptureError):
    """Invalid device"""
    def __init__(self, reason: str):
        super().__init__(f"Invalid device: {reason}")
        self.reason = reason
    
    @property
    def help_anchor(self) -> str:
        return "devices"


class DeviceDisconnectedError(AudioCaptureError):
    """Device was disconnected"""
    def __init__(self, device_name: str):
        super().__init__(f"Device '{device_name}' was disconnected")
        self.device_name = device_name
    
    @property
    def recovery_suggestion(self) -> str:
        return "Reconnect the audio device or select a different device"
    
    @property
    def help_anchor(self) -> str:
        return "devices"


class DeviceInUseError(AudioCaptureError):
    """Device is already in use"""
    def __init__(self, device_name: str):
        super().__init__(f"Device '{device_name}' is already in use")
        self.device_name = device_name
    
    @property
    def help_anchor(self) -> str:
        return "devices"


# Permission Errors
class PermissionDeniedError(AudioCaptureError):
    """Permission denied"""
    def __init__(self):
        super().__init__("Permission denied")


class AudioLoopbackPermissionError(AudioCaptureError):
    """System audio loopback permission required"""
    def __init__(self, message=None):
        if message:
            super().__init__(message)
        else:
            super().__init__(
                "System audio loopback permission is required. "
                "This feature requires Windows 10 version 1803 or later"
            )
    
    @property
    def failure_reason(self) -> str:
        return "The app needs access to system audio"
    
    @property
    def recovery_suggestion(self) -> str:
        return "Ensure you're running Windows 10 version 1803 or later and have appropriate permissions"
    
    @property
    def help_anchor(self) -> str:
        return "permissions"


class MicrophonePermissionError(AudioCaptureError):
    """Microphone permission required"""
    def __init__(self):
        super().__init__(
            "Microphone permission is required. "
            "Please grant permission in Windows Settings > Privacy > Microphone"
        )
    
    @property
    def failure_reason(self) -> str:
        return "The app needs access to microphone input"
    
    @property
    def recovery_suggestion(self) -> str:
        return "Open Windows Settings and grant Microphone permission to this app"
    
    @property
    def help_anchor(self) -> str:
        return "permissions"


# Session Errors
class SessionNotFoundError(AudioCaptureError):
    """Session not found"""
    def __init__(self, session_id: uuid.UUID):
        super().__init__(f"Session with ID {session_id} not found")
        self.session_id = session_id


class InvalidStateError(AudioCaptureError):
    """Invalid session state"""
    def __init__(self, message: str):
        super().__init__(f"Invalid session state: {message}")


class SessionAlreadyActiveError(AudioCaptureError):
    """Session is already active"""
    def __init__(self):
        super().__init__("Session is already active")


class SessionNotActiveError(AudioCaptureError):
    """Session is not active"""
    def __init__(self):
        super().__init__("Session is not active")


class SessionStartFailedError(AudioCaptureError):
    """Failed to start session"""
    def __init__(self, reason: str):
        super().__init__(f"Failed to start session: {reason}")
        self.reason = reason


# Format Errors
class UnsupportedFormatError(AudioCaptureError):
    """Unsupported audio format"""
    def __init__(self, format_desc: str):
        super().__init__(f"Unsupported audio format: {format_desc}")
        self.format_desc = format_desc
    
    @property
    def help_anchor(self) -> str:
        return "formats"


class FormatConversionFailedError(AudioCaptureError):
    """Format conversion failed"""
    def __init__(self, reason: str):
        super().__init__(f"Format conversion failed: {reason}")
        self.reason = reason
    
    @property
    def help_anchor(self) -> str:
        return "formats"


class FormatMismatchError(AudioCaptureError):
    """Audio format mismatch"""
    def __init__(self, details: str):
        super().__init__(f"Audio format mismatch: {details}")
        self.details = details
    
    @property
    def recovery_suggestion(self) -> str:
        return "Check that all audio components are using compatible formats"
    
    @property
    def help_anchor(self) -> str:
        return "formats"


class FormatNegotiationFailedError(AudioCaptureError):
    """Failed to negotiate compatible audio format"""
    def __init__(self):
        super().__init__("Failed to negotiate compatible audio format")
    
    @property
    def help_anchor(self) -> str:
        return "formats"


# Output Errors
class OutputNotConfiguredError(AudioCaptureError):
    """Output is not configured"""
    def __init__(self):
        super().__init__("Output is not configured")


class OutputConfigurationFailedError(AudioCaptureError):
    """Output configuration failed"""
    def __init__(self, reason: str):
        super().__init__(f"Output configuration failed: {reason}")
        self.reason = reason


class OutputProcessingFailedError(AudioCaptureError):
    """Output processing failed"""
    def __init__(self, reason: str):
        super().__init__(f"Output processing failed: {reason}")
        self.reason = reason


class FileWriteFailedError(AudioCaptureError):
    """File write failed"""
    def __init__(self, reason: str):
        super().__init__(f"File write failed: {reason}")
        self.reason = reason


class StreamingFailedError(AudioCaptureError):
    """Streaming failed"""
    def __init__(self, reason: str):
        super().__init__(f"Streaming failed: {reason}")
        self.reason = reason


# Buffer Errors
class BufferAllocationFailedError(AudioCaptureError):
    """Failed to allocate audio buffer"""
    def __init__(self):
        super().__init__("Failed to allocate audio buffer")


class BufferOverflowError(AudioCaptureError):
    """Audio buffer overflow"""
    def __init__(self):
        super().__init__("Audio buffer overflow")
    
    @property
    def failure_reason(self) -> str:
        return "Audio processing can't keep up with input rate"
    
    @property
    def recovery_suggestion(self) -> str:
        return "Try reducing the audio quality or closing other applications"
    
    @property
    def help_anchor(self) -> str:
        return "performance"


class BufferUnderrunError(AudioCaptureError):
    """Audio buffer underrun"""
    def __init__(self):
        super().__init__("Audio buffer underrun")
    
    @property
    def failure_reason(self) -> str:
        return "Audio input is not providing data fast enough"
    
    @property
    def help_anchor(self) -> str:
        return "performance"


class InvalidBufferSizeError(AudioCaptureError):
    """Invalid buffer size"""
    def __init__(self):
        super().__init__("Invalid buffer size")


# System Errors
class SystemResourcesExhaustedError(AudioCaptureError):
    """System resources exhausted"""
    def __init__(self):
        super().__init__("System resources exhausted")
    
    @property
    def failure_reason(self) -> str:
        return "Not enough CPU or memory available"
    
    @property
    def recovery_suggestion(self) -> str:
        return "Close other applications to free up system resources"


class MemoryAllocationFailedError(AudioCaptureError):
    """Memory allocation failed"""
    def __init__(self):
        super().__init__("Memory allocation failed")


class AudioEngineStartFailedError(AudioCaptureError):
    """Audio engine failed to start"""
    def __init__(self, reason: str):
        super().__init__(f"Audio engine failed to start: {reason}")
        self.reason = reason


class UnknownError(AudioCaptureError):
    """Unknown error"""
    def __init__(self, message: str):
        super().__init__(f"Unknown error: {message}")


# Network Errors
class NetworkConnectionFailedError(AudioCaptureError):
    """Network connection failed"""
    def __init__(self, reason: str):
        super().__init__(f"Network connection failed: {reason}")
        self.reason = reason


class StreamingProtocolError(AudioCaptureError):
    """Streaming protocol error"""
    def __init__(self, reason: str):
        super().__init__(f"Streaming protocol error: {reason}")
        self.reason = reason


# Windows-specific Errors
class WASAPIError(AudioCaptureError):
    """Windows Audio Session API error"""
    def __init__(self, error_code: int, message: str):
        super().__init__(f"WASAPI error {error_code}: {message}")
        self.error_code = error_code


class MMEError(AudioCaptureError):
    """Windows Multimedia Extension error"""
    def __init__(self, error_code: int, message: str):
        super().__init__(f"MME error {error_code}: {message}")
        self.error_code = error_code


class DirectSoundError(AudioCaptureError):
    """DirectSound error"""
    def __init__(self, error_code: int, message: str):
        super().__init__(f"DirectSound error {error_code}: {message}")
        self.error_code = error_code


class ErrorHandler:
    """Error handler with recovery strategies"""
    
    @staticmethod
    async def handle(
        error: Exception,
        strategy: ErrorRecoveryStrategy = ErrorRecoveryStrategy.FAIL,
        max_attempts: int = 3,
        delay: float = 1.0,
        fallback_action: Optional[Callable] = None
    ):
        """Handle error with recovery strategy"""
        print(f"Error occurred: {error}")
        
        if strategy == ErrorRecoveryStrategy.RETRY:
            # Retry logic would be implemented by caller
            raise error
            
        elif strategy == ErrorRecoveryStrategy.FALLBACK:
            if fallback_action:
                await fallback_action()
            else:
                raise error
                
        elif strategy == ErrorRecoveryStrategy.IGNORE:
            # Log and continue
            print(f"Ignoring error: {error}")
            
        elif strategy == ErrorRecoveryStrategy.FAIL:
            raise error
    
    @staticmethod
    def suggested_strategy(error: Exception) -> tuple[ErrorRecoveryStrategy, dict]:
        """Get suggested recovery strategy for error"""
        if isinstance(error, DeviceDisconnectedError):
            return ErrorRecoveryStrategy.RETRY, {"max_attempts": 3, "delay": 1.0}
        elif isinstance(error, (BufferOverflowError, BufferUnderrunError)):
            return ErrorRecoveryStrategy.IGNORE, {}
        elif isinstance(error, SessionStartFailedError):
            return ErrorRecoveryStrategy.RETRY, {"max_attempts": 2, "delay": 0.5}
        else:
            return ErrorRecoveryStrategy.FAIL, {}


class ErrorContext:
    """Error context for detailed debugging"""
    
    def __init__(
        self,
        error: Exception,
        session_id: Optional[uuid.UUID] = None,
        operation: str = "",
        additional_info: Optional[Dict[str, Any]] = None
    ):
        self.error = error
        self.timestamp = datetime.now()
        self.session_id = session_id
        self.operation = operation
        self.additional_info = additional_info or {}
    
    def report(self) -> str:
        """Create detailed error report"""
        report = f"""Audio Capture Error Report
========================
Timestamp: {self.timestamp}
Operation: {self.operation}
Error: {self.error}
"""
        
        if self.session_id:
            report += f"\nSession ID: {self.session_id}"
        
        if self.additional_info:
            report += "\n\nAdditional Information:"
            for key, value in self.additional_info.items():
                report += f"\n  {key}: {value}"
        
        if isinstance(self.error, AudioCaptureError):
            if self.error.recovery_suggestion:
                report += f"\n\nRecovery Suggestion: {self.error.recovery_suggestion}"
        
        return report